### ISSUE TITLE
Implement existing snippet detection for Tag Manager

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#433]
## Feature Description

We currently detect existing snippets for Analytics and AdSense. This detection should happen for Tag Manager as well, and cover non-AMP as well as AMP.

If an existing tag is found, it should work like in Analytics: If the user has access to that very container, it should be forced. If not, the user should be blocked from proceeding. In both cases, the user should be informed about the existing tag and that it comes from another code source.

In other words, this issue will also include a couple of UI changes/requirements.

This was originally reported as part of #413.

~~**This issue is blocked by #83 and #432, but theoretically implementation can at least start before these are fully completed.**~~

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When the user clicks "connect service" for GTM, Site Kit should confirm whether there's an existing GTM snippet in the FE (placed directly by the user or via another plugin). 
* Assuming is an existing snippet, check if the user's Google account has access to that GTM container (as an Admin https://support.google.com/tagmanager/answer/6107011?hl=en#account_permissions) 
* Display two radio buttons in GTM settings for whether the snippet should be placed or not (`useSnippet`); in GTM setup this should not be displayed and instead automatically determined:
    * if there is an existing snippet on the page, auto-set this to false
    * if there is no existing snippet on the page, auto-set this to true 

Case 1: the account has access to the container:
* Force-select these and give user option to proceed. 
* Messaging in the UI should be similar to GA (_"An existing tag was found on your site with the id XYZ. If later on you decide to replace this tag, Site Kit can place the new tag for you. Make sure you remove the old tag first."_)

Case 2: the account doesn't have access to the container:
* block the user from proceeding (give "back" option)
* display a message that the user either has to get access as Admin for the container, or remove the old container in order to place a new one: _"We've detected there's already an existing Tag Manager tag on your site (ID %s), but your account doesn't seem to have access to this Tag Manager container. You can either remove the existing tag and connect to a different account, or request access to this container from your team."_

## Implementation Brief

* Update `getExistingTag` to support Tag Manager snippet pattern matching for `GTM-*` container public ID
    - Snippets may be able to be included in different ways so be sure to check for different types of JS code that include GTM. It's worth checking for the [official code](https://developers.google.com/tag-manager/quickstart) as well as GTM code supplied by different, popular WordPress plugins such as [Monster Insights](https://www.monsterinsights.com) that enable this functionality. Checking for the AMP and non-AMP versions should also be done.
    - Suppress `Tag_Manager` tag output for `$_GET[tagverify]` requests (see Analytics and AdSense modules)
* Update Tag Manager setup to use `getExistingTag`
    - If detected, the found container ID is matched against available container IDs that the current user has access to  
    _The user must have `publish` permission for the individual container_ 
        - If found and the user has the required permission, pre-select this container and disable the dropdown  
        _See messaging in AC_
        - If found and the does not have the required permission, display error message (see AC) and display "Back" button for user to abort setup.
    - If not detected, attempt to pre-select a container by matching the container name using the same name generated by `Tag_Manager::create_container()` (#361)
    - If not detected and no matching container, list accounts and containers with none selected
    - Set `useSnippet` setting based on presence of existing tag (if existing tag, `useSnippet = false`, otherwise default to `true`) 
* Update Tag Manager settings component to expose the new `useSnippet` setting toggle (most of this can be borrowed from Analytics)
* Add filter to Tag Manager settings option which defaults `useSnippet` to `true` if the key does not exist yet

## Changelog entry

* Detect already existing Tag Manager snippets from other sources and inform about them in the setup flow.

**Comment by felixarntz** [comment]
@tofumatt 
> If multiple snippets are found, it would be worth notifying the user that multiple snippets are already in the page and that only one should be used. (There should only be one GTM snippet on the page, correct?)

This is out of scope for this issue, we also don't consider multiple in the other modules. Worth discussing for the future though as a general enhancement.

> Similar to how Google Analytics snippets are detected

You don't necessarily need to add this in detail as part of the IB, but just to flag it, it needs careful investigation in which ways Tag Manager can be included in a page, including AMP and non-AMP versions. The regexes used need to provide sufficient coverage of different alternatives. At least a note on this might be good to have here.

**Comment by tofumatt** [comment]
I've updated the IB to remove the multiple properties part and added the note about being diligent about checking for different Tag Manager include snippets in our checks.

**Comment by felixarntz** [comment]
IB ✅

**Comment by aaemnnosttv** [comment]
> Assuming is an existing snippet, check if the user's Google account has access to that GTM container (as an Admin https://support.google.com/tagmanager/answer/6107011?hl=en#account_permissions)

"User" or "Admin" permissions are at the account level. Containers can have individual permissions apart from that.

Do we want to specifically check that the user has Admin level permissions on the container's account or should we instead check for "full" permissions on the container itself (i.e "publish")?

Note: we can check this using the `user_permissions` API endpoint.
https://developers.google.com/tag-manager/api/v2/reference/accounts/user_permissions/list?authuser=1

**Comment by marrrmarrr** [comment]
@aaemnnosttv agreed - better to be more granular here and focus on individual container permissions rather than account-level permissions.

**Comment by aaemnnosttv** [comment]
Thanks @marrrmarrr this is actually fairly straightforward as Admins inherit the `publish` permission on each container so we only need to check for that; no need to do any additional checks as to whether it is a direct or inherited permission.

**Comment by aaemnnosttv** [comment]
@felixarntz @marrrmarrr it seems that much of the acceptance criteria has been borrowed from Analytics but I think there are some important differences to consider here which may not have been evident before.

Perhaps the most fundamental difference is that Analytics also brings data into WordPress, but Tag Manager is currently a simple client-side only integration (outputting the snippet) with no data in WordPress other than account and container names to choose from.

In the case of an existing tag, I don't think we necessarily need to block the user from completing setup or check their permissions. Compared to Analytics, all we're doing is outputting the script on the front end which loads tags from the container. I don't think there is any reason why we can't output more than one (not the case of course with Analytics).

Analytics has the concept (and setting for) whether or not to output the snippet because data can still be shown in the plugin while something else is responsible for outputting the tag which generates that data. This is not the case with Tag Manager as its sole responsibility is outputting a script to load the container. In this case, it doesn't make sense to force the setup to use the account and container IDs from an existing tag because it would simply result in duplicate tags on the front.

For these reasons I think that permissions/access/messaging around existing tags could be greatly simplified/removed with more of a focus on making the user aware, and only block them (perhaps) from setting the site up with the _same tag_ as an existing tag to prevent duplicates.

**Comment by felixarntz** [comment]
@aaemnnosttv Good point. I agree we need to ensure there are no duplicate tags in the frontend. However, similar to Analytics we also need to ensure there are not two Tag Manager tags in the frontend, that's why we need this check for existing tag.

In Analytics it's actually the same, it doesn't technically hurt to have more than one tracking snippet, it would just track the same thing to different Analytics properties, and only the data from the one active in Site Kit would show in the plugin. However, we were advised by the Analytics team that this should be avoided.

Summarizing, I think we should for now block users without access the same way we do for Analytics to prevent multiple Tag Manager tags in the frontend. In the case of an existing tag, we then need to set a flag in the database that ensures our own snippet is not printed. I'd suggest just using another `useSnippet` setting (specific to Tag Manager), which however is not exposed in the UI at all, just auto-populated based on whether there's an existing tag (--> `false`) or not (--> `true`).

**Comment by aaemnnosttv** [comment]
> I think we should for now block users without access the same way we do for Analytics to prevent multiple Tag Manager tags in the frontend.

I'm not entirely sure what you mean here. What does the user's access have to do with preventing multiple tags? If we want to prevent multiple then completing setup shouldn't be possible if there is an existing tag.

> In the case of an existing tag, we then need to set a flag in the database that ensures our own snippet is not printed.

Why allow the user to complete the setup if it won't output the tag? This sounds like a confusing experience for the user. If we were to set `useSnippet` at setup time to `false` and not output the tag, what would cause it to change to `true` if there is no UI?

I think we should simply block setup from completing if there is any existing tag if we don't want to allow for multiple.

**Comment by felixarntz** [comment]
> If we want to prevent multiple then completing setup shouldn't be possible if there is an existing tag.

Yes, but we'd only want to block if the user doesn't have access. If the user _has_ access, we can consider Tag Manager active, force-set the account that has access to the existing tag, and just not print it ourselves (hence the `useSnippet`: `false`). That is similar to how we do it in Analytics.

> Why allow the user to complete the setup if it won't output the tag? This sounds like a confusing experience for the user.

It's the same for Analytics. We allow users to complete the setup if there is an existing tag they have access to. Under those circumstances, we force-choose that property and don't place the tag ourselves. I think for GTM we can do the same. When you access the module settings, we should re-check for the existing tag, and if there's none anymore, we can silently set `useSnippet` to `true` and then also allow the user to change the container.

**Comment by aaemnnosttv** [comment]
> Yes, but we'd only want to block if the user doesn't have access. 

I'm still not quite understanding why the access matters with an existing tag. Whether the user has access or not, the tag is still on the site and the Tag Manager module should not add an additional tag; the behavior is the same either way. 

I understand this is how we do things in Analytics and there are good reasons for that but I don't think the same applies for Tag Manager.

> If the user has access, we can consider Tag Manager active, force-set the account that has access to the existing tag, and just not print it ourselves (hence the useSnippet: false). That is similar to how we do it in Analytics.

We could do that, but is this not adding extra code/effort/maintenance to support the module being activated in a way that does nothing? If Tag Manager module's sole purpose is to output its snippet and nothing else, why would we allow it to be activated if it would not do that?

Not trying to be overly difficult here, just trying to understand the reasoning and make sure we land on the best experience for the user.

**Comment by aaemnnosttv** [comment]
@felixarntz Moving this to blocked for now per our conversation to focus on higher priority items.

**Comment by felixarntz** [comment]
@aaemnnosttv 
@marrrmarrr and I revisited this, let's do the implementation the same way we do for Analytics:
* If there is an existing snippet and the user does not have access, block them.
* If they do have access to them, force-select it and do not place our own.

The reasoning is the following: We'll expand what the GTM module does, and even if it technically does nothing at this point, we should allow the user to complete the setup.

Note that we'll also want a setting for whether the snippet is placed, that is initially auto-determined, but the user can adjust it in settings afterwards (again similar to Analytics). I've updated the ACs, sending this back to IB now so that this part can be added.

**Comment by shirokoweb** [comment]
Found about this topic too late... after noticing GA was added twice. Now I understand why, but reading topic gave me some serious headache.
 
I think you guys over-complicating things with endless conditions / possibilities (oh and what about this, and what about that, and what if...) **where a simple**:
"Would you like GSK to add GA snippet for You ? YES / NO"
would save you much time and give more control to end users.

We for example, load GA through GTM, but, GTM has to be allowed first by visitor (GDPR consent).
So if our visitor doesn't give consent for GTM, GA won't be loaded either, nor GSK will detect it's implemented. Therefore, GSK will add whatever tracking it "has" to add on our behalf...

@aaemnnosttv well, in countries where GDPR is taken very seriously, **collect explicit consent prior adding any cookie or tracking code is a priority**.

**Comment by felixarntz** [comment]
IB ✅

**Comment by cole10up** [comment]
**Tested**
Pasted GTM snippet in header and saved
![image](https://user-images.githubusercontent.com/57115907/72410007-32da4a00-371c-11ea-9500-ce4ae198b3f3.png)

Verified snippet detection occurred:
![image](https://user-images.githubusercontent.com/57115907/72410101-6e751400-371c-11ea-9bb1-6a216ec981e7.png)

Confirmed cancel button functions.

Verified permissions message displays properly when user doesn't have the correct permissions:
Removed permissions
![image](https://user-images.githubusercontent.com/57115907/72410494-4afe9900-371d-11ea-9eb7-6c4793a74240.png)

![image](https://user-images.githubusercontent.com/57115907/72410525-5f429600-371d-11ea-9dcb-6d56c40077fb.png)

Picky but @tofumatt, should the above 'Cancel' button be labeled 'Back'?

Passed QA ✅

**Comment by aaemnnosttv** [comment]
> Picky but @tofumatt, should the above 'Cancel' button be labeled 'Back'?

This is consistent with the other setup modules. Also I think it may not be possible to have different wording for the cancel button currently due the way the setup module components are composed.

**Comment by tofumatt** [comment]
It's definitely not nice UX though, agreed. We should fix those en-masse—could you create an issue for any modules that should have that wording changed @cole10up?

**Comment by marrrmarrr** [comment]
LGTM. 
Also SGTM on the s/cancel/back/g for cases where the user hasn't really done any action to cancel, like the one Cole described above.

**Comment by cole10up** [comment]
@tofumatt @marrrmarrr I've created this ticket to adjust verbiage where user's presented an expected error message: https://github.com/google/site-kit-wp/issues/1045

I'm unsure of other cases, please advise.
